modules/javafx.base/src/main/java/javafx/beans/property/Property.java

```diff
@@ -40,76 +40,97 @@
+     * established by calling {@link #bind(ObservableValue)}.
+     * <p>
+     * Note that this method does <b>not</b> account for bidirectional bindings that were
+     * established by calling {@link #bindBidirectional(Property)}.
      *
-     * @return {@code true} if the {@code Property} is bound, {@code false}
-     *         otherwise
+     * @return whether this property is unidirectionally bound
      */
     boolean isBound();
 
     /**
-     * Create a bidirectional binding between this {@code Property} and another
-     * one.
-     * Bidirectional bindings exists independently of unidirectional bindings. So it is possible to
-     * add unidirectional binding to a property with bidirectional binding and vice-versa. However, this practice is
-     * discouraged.
+     * Establishes a bidirectional binding between this property and another {@link Property}.
+     * <p>
+     * After establishing the binding, the values of both properties are synchronized: any change
+     * to the value of one property will immediately result in the value of the other property being
+     * changed accordingly.
      * <p>
-     * It is possible to have multiple bidirectional bindings of one Property.
+     * While it is not possible for a property to be bound by more than one unidirectional binding,
+     * it is legal to establish multiple bidirectional bindings for the same property. However,
+     * since a bidirectional binding allows for the values of both properties to be changed
+     * by calling {@link #setValue(Object)}, neither of the properties is considered to be a
+     * <i>bound property</i> that returns {@code true} from {@link #isBound()}.
      * <p>
-     * JavaFX bidirectional binding implementation use weak listeners. This means bidirectional binding does not prevent
-     * properties from being garbage collected.
+     * Both properties of a bidirectional binding <i>weakly</i> reference their counterparts.
+     * This is different from a unidirectional binding, where the bound property <i>strongly</i>
+     * references its binding source. In practice, this means that if any of the bidirectionally
+     * bound properties become unreachable, the binding is eligible for garbage collection.
+     * Furthermore, neither of the bidirectionally bound properties will keep its counterpart
+     * alive if the counterpart would otherwise become unreachable.
+     * <p>
+     * Bidirectional bindings are independent from unidirectional bindings. As a consequence,
+     * establishing a bidirectional binding does not remove a unidirectional binding that has
+     * already been established, nor does it prevent a unidirectional binding from being
+     * established. However, doing so is not a meaningful use of this API, because the
+     * unidirectionally bound property will cause any attempt to change the value of any of
+     * the properties to fail with an exception.
```

Comment:
Maybe we could simply the mental model of the property specification by making it illegal in all cases to use unidirectional and bidirectional bindings at the same time. The specification would be reduced to "it's illegal", instead of having a long explanation that says it's only _likely_ to cause an exception.

Most attempts to use both kinds of bindings will fail with an exception anyway. Some will fail instantly when calling `bindBidirectional`, some will fail later when calling `setValue`. Some will fail more silently because the exception will happen inside `BidirectionalBinding` and not bring down the application. Only a kind-of-exotic situation will not produce exceptions at all (the situation that you described in the second bullet point).

If we were to always fail instantly at the point of calling `bind` or `bindBidirectional`, we would make the misuse of this API visible where it's most relevant, instead of failing later in other parts of the codebase.

Reply:
The following unit test demonstrated the *current* behavior, though maybe it's not the desired one:

```java
import static org.junit.Assert.*;

import javafx.beans.property.IntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;

import org.junit.Before;
import org.junit.Test;

public class BindingsTest {

	private IntegerProperty property1, property2, property3;

	@Before
	public void setup() {
		property1 = new SimpleIntegerProperty(1);
		property2 = new SimpleIntegerProperty(2);
		property3 = new SimpleIntegerProperty(3);
	}

	@Test
	public void testBidiAndThenUni() {
		property1.bindBidirectional(property2);
		assertEquals(2, property1.get());
		assertEquals(2, property2.get());
		property1.bind(property3);
		assertEquals(3, property1.get());
		assertEquals(3, property2.get());
		assertEquals(3, property3.get());
	}

	@Test(expected = RuntimeException.class)
	public void setProperty1() {
		property1.bindBidirectional(property2);
		property1.bind(property3);

		property1.set(4);
	}

	@Test
	public void setProperty2() {
		property1.bindBidirectional(property2);
		property1.bind(property3);

		// ignores exception (but still prints stack trace) at
		// com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:181)
		property2.set(5);
		assertEquals(3, property1.get());
		assertEquals(3, property2.get()); // "Bidirectional binding failed, setting to the previous value"
		assertEquals(3, property3.get());
	}

	@Test
	public void setProperty3() {
		property1.bindBidirectional(property2);
		property1.bind(property3);

		property3.set(6);
		assertEquals(6, property1.get());
		assertEquals(6, property2.get());
		assertEquals(6, property3.get());
	}

	@Test(expected = RuntimeException.class)
	public void testUniAndThenBidiOnBound() {
		property1.bind(property3);
		property1.bindBidirectional(property2);
	}

	@Test
	public void testUniAndThenBidiOnUnbound() {
		property1.bind(property3);
		property2.bindBidirectional(property1);
		
		assertEquals(3, property1.get());
		assertEquals(3, property2.get());
		assertEquals(3, property3.get());
	}
}
```

My observations:

* `setProperty2()` is weird in that it throws an exception internally which is ignored by the JVM, and then reverts the change to the unidirectionally unbound property when it finds out that its counterpart can't be changed. It makes sense logically, but I don't see a valid use case for this.
* `setProperty3()` behaves as I expect it to, but this is effectively 2 chained unidirectional binds: `property3` changes `property1` which changes `property2`. This is the same as `testUniAndThenBidiOnUnbound()`.

It would make sense to disallow this mixing of bindings on grounds that this is probably not what the developer wants to do. If we opt to let them do it but write that it should not be done, we could emit a warning. I'm leaning towards to fail-fast approach.

Reply:
I also like documenting it as illegal, but am hesitant to make any implementation change in this area for JavaFX 17 during rampdown. So I would recommend documenting as illegal, but not adding the `@throws` to the docs. We could still say that an exception might be thrown or could say something like "results are undefined". Then in JavaFX 18, we can add the `@throws` and change the implementation to fail fast.

Reply:
Fine with me.

I didn't look at the current tests, but if it helps, we can add the one I wrote.